server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1463,4 +2157,171 @@ protected ClusterBlockException checkBlock(UpdateIndexShardSnapshotStatusRequest
+     */
+    private final class FailPendingRepoTasksTask extends ClusterStateUpdateTask {
+
+        // Snapshots to fail after the state update
+        private final List<Snapshot> snapshotsToFail = new ArrayList<>();
+
+        // Delete uuids to fail because after the state update
+        private final List<String> deletionsToFail = new ArrayList<>();
+
+        // Failure that caused the decision to fail all snapshots and deletes for a repo
+        private final Exception failure;
+
+        private final String repository;
+
+        FailPendingRepoTasksTask(String repository, Exception failure) {
+            this.repository = repository;
+            this.failure = failure;
+        }
+
+        @Override
+        public ClusterState execute(ClusterState currentState) {
+            final SnapshotDeletionsInProgress deletionsInProgress =
+                    currentState.custom(SnapshotDeletionsInProgress.TYPE, SnapshotDeletionsInProgress.EMPTY);
+            boolean changed = false;
+            final List<SnapshotDeletionsInProgress.Entry> remainingEntries = deletionsInProgress.getEntries();
+            List<SnapshotDeletionsInProgress.Entry> updatedEntries = new ArrayList<>(remainingEntries.size());
+            for (SnapshotDeletionsInProgress.Entry entry : remainingEntries) {
+                if (entry.repository().equals(repository)) {
+                    changed = true;
+                    deletionsToFail.add(entry.uuid());
+                } else {
+                    updatedEntries.add(entry);
+                }
+            }
+            final SnapshotDeletionsInProgress updatedDeletions = changed ? SnapshotDeletionsInProgress.of(updatedEntries) : null;
+            final SnapshotsInProgress snapshotsInProgress =
+                currentState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY);
+            final List<SnapshotsInProgress.Entry> snapshotEntries = new ArrayList<>();
+            boolean changedSnapshots = false;
+            for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
+                if (entry.repository().equals(repository)) {
+                    // We failed to read repository data for this delete, it is not the job of SnapshotsService to
+                    // retry these kinds of issues so we fail all the pending snapshots
+                    snapshotsToFail.add(entry.snapshot());
+                    changedSnapshots = true;
+                } else {
+                    // Entry is for another repository we just keep it as is
+                    snapshotEntries.add(entry);
+                }
+            }
+            final SnapshotsInProgress updatedSnapshotsInProgress = changedSnapshots ? SnapshotsInProgress.of(snapshotEntries) : null;
+            return updateWithSnapshots(currentState, updatedSnapshotsInProgress, updatedDeletions);
+        }
+
+        @Override
+        public void onFailure(String source, Exception e) {
+            logger.debug(
```

Comment:
should we log at info level? Important event, no?

Reply:
Yea we could log this more visibly. It's just something that will happen on master-failover and I didn't want to addd to the general noise of that but that's not an actual argument => info it is :)